pkg/kubelet/pod_workers.go

```diff
@@ -221,76 +602,412 @@ func (p *podWorkers) UpdatePod(options *UpdatePodOptions) {
+// clears the working status.
+func (p *podWorkers) completeWorkQueueNext(uid types.UID) {
 	p.podLock.Lock()
 	defer p.podLock.Unlock()
 	if workUpdate, exists := p.lastUndeliveredWorkUpdate[uid]; exists {
 		p.podUpdates[uid] <- workUpdate
 		delete(p.lastUndeliveredWorkUpdate, uid)
 	} else {
-		// put this line in removeWorker may cause dead lock, so keep reset it here
-		delete(p.isWorking, uid)
+		p.podSyncStatuses[uid].working = false
 	}
 }
 
+// contextForWorker returns or initializes the appropriate context for a known
+// worker. If the current context is expired, it is reset. If no worker is
+// present, no context is returned.
+func (p *podWorkers) contextForWorker(uid types.UID) context.Context {
+	p.podLock.Lock()
+	defer p.podLock.Unlock()
+
+	status, ok := p.podSyncStatuses[uid]
+	if !ok {
+		return nil
+	}
+	if status.ctx == nil || status.ctx.Err() == context.Canceled {
+		status.ctx, status.cancelFn = context.WithCancel(context.Background())
+	}
+	return status.ctx
+}
+
+// SyncKnownPods will purge any fully terminated pods that are not in the desiredPods
+// list, which means SyncKnownPods must be called in a threadsafe manner from calls
+// to UpdatePods for new pods. It returns a map of known workers that are not finished
+// with a value of SyncPodTerminated, SyncPodKill, or SyncPodSync depending on whether
+// the pod is terminated, terminating, or syncing.
+func (p *podWorkers) SyncKnownPods(desiredPods []*v1.Pod) map[types.UID]PodWorkType {
+	workers := make(map[types.UID]PodWorkType)
+	known := make(map[types.UID]struct{})
+	for _, pod := range desiredPods {
+		known[pod.UID] = struct{}{}
+	}
+
+	p.podLock.Lock()
+	defer p.podLock.Unlock()
+
+	p.podsSynced = true
+	for uid, status := range p.podSyncStatuses {
+		if _, exists := known[uid]; !exists {
+			p.removeTerminatedWorker(uid)
+		}
+		switch {
+		case !status.terminatedAt.IsZero():
+			workers[uid] = TerminatedPodWork
+		case !status.terminatingAt.IsZero():
+			workers[uid] = TerminatingPodWork
+		default:
+			workers[uid] = SyncPodWork
+		}
+	}
+	return workers
+}
+
+// removeTerminatedWorker cleans up and removes the worker status for a worker that
+// has reached a terminal state of "finished" - has successfully exited
+// syncTerminatedPod. This "forgets" a pod by UID and allows another pod to be recreated
+// with the same UID.
+func (p *podWorkers) removeTerminatedWorker(uid types.UID) {
+	status, ok := p.podSyncStatuses[uid]
+	if !ok {
+		// already forgotten, or forgotten too early
+		klog.V(4).InfoS("Pod worker has been requested for removal but is not a known pod", "podUID", uid)
+		return
+	}
+
+	if !status.finished {
+		klog.V(4).InfoS("Pod worker has been requested for removal but is still not fully terminated", "podUID", uid)
+		return
+	}
+
+	klog.V(4).InfoS("Pod has been terminated and is no longer known to the kubelet, remove all history", "podUID", uid)
+	delete(p.podSyncStatuses, uid)
```

Comment:
For static pod,  pod.uid does not change, while pod REMOVE and ADD encountes race condition,  `delete(p.podSyncStatuses, uid)` here may cause deadlock,  as the old podwoker may write workUpdate to new podworker's podUpdate channel  in `completeWorkQueueNext(uid types.UID) `

Reply:
Thanks, we don't have a test case for this in node tests, which I'll add.

Reply:
Turns out we have bugs in this area - a pod created like this doesn't seem to have its mirror pod updated.  Digging into why that happens before I finish up the area here.

Reply:
https://github.com/kubernetes/kubernetes/pull/103374 is a test that fails in master (indicating the mirror pod not getting updated).

Reply:
> For static pod, pod.uid does not change, while pod REMOVE and ADD encountes race condition, delete(p.podSyncStatuses, uid) here may cause deadlock, as the old podwoker may write workUpdate to new podworker's podUpdate channel in completeWorkQueueNext(uid types.UID)

Can you elaborate more about how the deadlock happens? IIUC, `SyncKnownPods` runs `HandlePodCleanups`, and during which no desired pod will be added/removed. In that case, if the static pod is seen, we keep the pod worker; if it is not seen, we remove the pod worker.

Why would `delete(p.podSyncStatuses, uid)` cause a deadlock?

Reply:
> > For static pod, pod.uid does not change, while pod REMOVE and ADD encountes race condition, delete(p.podSyncStatuses, uid) here may cause deadlock, as the old podwoker may write workUpdate to new podworker's podUpdate channel in completeWorkQueueNext(uid types.UID)
> 
> Can you elaborate more about how the deadlock happens? IIUC, `SyncKnownPods` runs `HandlePodCleanups`, and during which no desired pod will be added/removed. In that case, if the static pod is seen, we keep the pod worker; if it is not seen, we remove the pod worker.
> 
> Why would `delete(p.podSyncStatuses, uid)` cause a deadlock?
> 
> But if a static is moved out of the manifest directory and then moved back, it is not quite clear to me what would happen. I feel like the termination will continue because the uid (generated from pod hash) is not changed.

@Random-Liu The case mybe like this below(We have encountered in v1.19 by add `delete(p.isWorking, uid)` here): 
1. For a static pod A terminated, `SyncKnownPods` calls `removeTerminatedWorker`, but at this time, the old podworker does not absolutely exited, it may have not run to the line `p.completeWorkQueueNext(pod.UID)`
2. Just this time. the static pod A manifest comes back, as  `removeTerminatedWorker` clears all the data like  `p.podUpdates[uid]`, so UpdatePod() will create a new podwoker. and the condtion below still satisfies, then the now podworker becomes running
```
if !status.IsWorking() {
		status.working = true
		podUpdates <- work
		return
	}
```
 so at this mini time, there are two podworders runing with some same shared data struct like `p.podUpdates[uid]` and `p.lastUndeliveredWorkUpdate[pod.UID]`
3. At this time, there are still some podUpdates come, UpdatePod() will put the podUpdate to  `p.lastUndeliveredWorkUpdate[pod.UID]`, just then , the old podworker run into the func `p.completeWorkQueueNext(pod.UID)`, and it writes a podUpdate to p.podUpdates[uid] channel and exits 
4. Just this time,  UpdatePod() will still put a podUpdate to  `p.lastUndeliveredWorkUpdate[pod.UID]`, just then , the new podworker run into the func `p.completeWorkQueueNext(pod.UID)`, as the p.podUpdates[uid] channel can only put one data. so it could not write successfully, then deadlock causes

Reply:
For 1.

The current code only actually `removeTerminatedWorker` when the pod worker `finished`. And the pod worker only `finished` at the end of `completeTerminated`, and the pod worker will return right after that.

I don't see how the case you described could happen.

Reply:
@Random-Liu Thanks for your ack.  I'm sorry for that I ignored the `managePodLoop  `will exit after `completeTerminated`, I mentioned the case frontend we have found in original code. Now this new change don't have this nit, I have no problem,